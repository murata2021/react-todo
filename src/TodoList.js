import React,{useState} from "react";
import NewTodoForm from "./NewTodoForm"
import Todo from "./Todo";
import uuid from 'react-uuid';



const TodoList=()=>{

    const INITIAL_TODOS_STATE= JSON.parse(localStorage.getItem("todos"))||[];

    // const INITIAL_TODOS_STATE= [];

    const [todos,setTodos]=useState(INITIAL_TODOS_STATE)

    const addTodo=async (newTodo)=>{

        newTodo.id=uuid()
        // console.log(newTodo)
        await setTodos(todos=> {

            console.log(todos)
            
            return [...todos,{...newTodo}]
        })
        // localStorage.setItem("todos", JSON.stringify(todos));
    }

    
    console.log(todos)

    const completed=(id)=>{
        const copyList=todos.map(todo=>{
            if (todo.id===id){
                todo.isCompleted=!todo.isCompleted
            }
            return todo
        })

        // console.log(copyList)
        setTodos(todos=>copyList)
        console.log(todos)
        // localStorage.setItem("todos", JSON.stringify(todos));

    }

    const deleteTodo=(id)=>{

        let copyList=[...todos.filter(todo=>todo.id!==id)]
        setTodos(todos=>copyList)

        // localStorage.setItem("todos", JSON.stringify(todos));

    }

    // console.log(todos)

    localStorage.setItem("todos", JSON.stringify(todos));

    


    return (
        <div>
            <h2>Things need to be done!</h2>
            <NewTodoForm addTodo={addTodo} />
            {todos.length>0?<h4>My Todo List</h4>:null}
            <ul>
                {todos.map(todo=><Todo id={todo.id} isCompleted={todo.isCompleted} key={todo.id} todo={todo.todo} completed={completed} deleteTodo={deleteTodo}/>)}
            </ul>
            
        </div>
    )

}

export default TodoList;