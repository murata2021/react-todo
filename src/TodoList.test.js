import { render, screen,fireEvent } from '@testing-library/react';
import TodoList from './TodoList';



it('renders without crashing', () => {
  render(<TodoList />);
});

it("matches snapshot", function() {
    const { asFragment } = render(<TodoList />);
    expect(asFragment()).toMatchSnapshot();
});


it("adds new todo",()=>{
    const {debug,queryByText,getByLabelText}=render(<TodoList/>);

    const btn=queryByText("Submit")
    let TodoInput=getByLabelText('Todo')
    

    
    expect(queryByText('❌')).not.toBeInTheDocument()


    fireEvent.change(TodoInput,{target:{value:'clean your room'}})
    fireEvent.click(btn)

    let deleteMe=queryByText('❌')

    expect(queryByText('❌')).toBeInTheDocument()

    console.log(deleteMe.closest('span'))
    expect(deleteMe.previousSibling.previousSibling).toHaveStyle(`
    text-decoration: None;
  `);

    fireEvent.click(deleteMe)
})

it("strikethroughs todo when it is clicked on",()=>{
    const {debug,queryByText,getByLabelText}=render(<TodoList/>);

    const btn=queryByText("Submit")
    let TodoInput=getByLabelText('Todo')
    

    fireEvent.change(TodoInput,{target:{value:'clean your room'}})
    fireEvent.click(btn)

    let deleteMe=queryByText('❌')

    expect(queryByText('❌')).toBeInTheDocument()

    expect(deleteMe.previousSibling.previousSibling).toHaveStyle(`
    text-decoration: None;
  `);

    let task=queryByText('clean your room')

    fireEvent.click(task)
    expect(task).toHaveStyle(`
    text-decoration: line-through;
    `);
    fireEvent.click(deleteMe)
})

it("deletes todo when the delete button is clicked on",()=>{
    const {debug,queryByText,getByLabelText}=render(<TodoList/>);

    const btn=queryByText("Submit")
    let TodoInput=getByLabelText('Todo')
    

    fireEvent.change(TodoInput,{target:{value:'clean your room'}})
    fireEvent.click(btn)
    let task=queryByText('clean your room')
    let deleteMe=queryByText('❌')

    expect(queryByText('❌')).toBeInTheDocument()

    fireEvent.click(deleteMe)
    expect(deleteMe).not.toBeInTheDocument()
    expect(task).not.toBeInTheDocument()
    
})