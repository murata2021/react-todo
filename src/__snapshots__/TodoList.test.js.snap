// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h2>
      Things need to be done!
    </h2>
    <div>
      <form>
        <label
          for="todo"
        >
          Todo 
        </label>
        <input
          id="todo"
          name="todo"
          placeholder="I need to do ..."
          type="text"
          value=""
        />
        <button>
          Submit
        </button>
      </form>
    </div>
    <ul />
  </div>
</DocumentFragment>
`;
